File: scripts/migrate-books.js
import mongoose from "mongoose";

import { connectToDatabase } from "../src/lib/db.js";
import { buildBookDocument } from "../src/lib/books.js";
import Book from "../src/models/Book.js";

async function ensureBook(details) {
  const bookDocument = buildBookDocument({
    title: details.title || "Unknown Book",
    author: details.author || "Unknown Author",
    level: details.level || "Unknown Level",
    totalPages: Number(details.totalPages) || Math.max(Number(details.pagesRead) || 1, 1),
  });

  return Book.findOneAndUpdate(
    { slug: bookDocument.slug },
    { $setOnInsert: bookDocument },
    { returnDocument: "after", upsert: true }
  );
}

async function migrate() {
  await connectToDatabase();

  const kidsCollection = mongoose.connection.db.collection("kids");
  const rawKids = await kidsCollection.find({}).toArray();
  let updatedKids = 0;

  for (const kid of rawKids) {
    const currentBooks = kid.reading?.currentBooks ?? [];
    const sessions = kid.reading?.sessions ?? [];
    const titleToBookId = new Map();
    let changed = false;

    const nextCurrentBooks = [];

    for (const currentBook of currentBooks) {
      if (currentBook.bookId) {
        nextCurrentBooks.push({
          bookId: currentBook.bookId,
          currentPage: currentBook.currentPage ?? 0,
        });
        continue;
      }

      const book = await ensureBook(currentBook);
      titleToBookId.set(currentBook.title, book._id.toString());
      nextCurrentBooks.push({
        bookId: book._id,
        currentPage: currentBook.currentPage ?? 0,
      });
      changed = true;
    }

    const nextSessions = [];

    for (const session of sessions) {
      if (session.bookId) {
        nextSessions.push(session);
        continue;
      }

      let bookId = titleToBookId.get(session.bookTitle);

      if (!bookId) {
        const matchingBook = currentBooks.find((book) => book.title === session.bookTitle);
        const book = await ensureBook(
          matchingBook || {
            title: session.bookTitle,
            pagesRead: session.pagesRead,
          }
        );

        bookId = book._id.toString();
      }

      nextSessions.push({
        ...session,
        bookId: new mongoose.Types.ObjectId(bookId),
      });
      changed = true;
    }

    if (!changed) {
      continue;
    }

    await kidsCollection.updateOne(
      { _id: kid._id },
      {
        $set: {
          "reading.currentBooks": nextCurrentBooks,
          "reading.sessions": nextSessions,
        },
      }
    );

    updatedKids += 1;
  }

  const bookCount = await Book.countDocuments();
  console.log(`Migrated ${updatedKids} kid documents. Books collection now has ${bookCount} records.`);
}

migrate()
  .then(() => process.exit(0))
  .catch((error) => {
    console.error("Failed to migrate books:", error);
    process.exit(1);
  });

File: scripts/migrate_books.py
from pathlib import Path
import sys

from bson import ObjectId
from pymongo import ReturnDocument


ROOT_DIR = Path(__file__).resolve().parents[1]
if str(ROOT_DIR) not in sys.path:
    sys.path.insert(0, str(ROOT_DIR))

from kid_quest.books import build_book_document
from kid_quest.db import ensure_indexes, get_db


def ensure_book(books_collection, details: dict) -> dict:
    book_document = build_book_document(
        details.get("title") or "Unknown Book",
        details.get("author") or "Unknown Author",
        details.get("level") or "Unknown Level",
        int(details.get("totalPages") or max(int(details.get("pagesRead") or 1), 1)),
    )

    return books_collection.find_one_and_update(
        {"slug": book_document["slug"]},
        {"$setOnInsert": book_document},
        upsert=True,
        return_document=ReturnDocument.AFTER,
    )


def normalize_object_id(value):
    if isinstance(value, ObjectId):
        return value

    if isinstance(value, str) and ObjectId.is_valid(value):
        return ObjectId(value)

    return None


def main():
    database = get_db()
    ensure_indexes()

    kids_collection = database["kids"]
    books_collection = database["books"]

    raw_kids = list(kids_collection.find({}))
    updated_kids = 0

    for kid in raw_kids:
        reading = kid.get("reading", {})
        current_books = reading.get("currentBooks", [])
        sessions = reading.get("sessions", [])
        title_to_book_id = {}
        changed = False

        next_current_books = []
        for current_book in current_books:
            current_book_id = normalize_object_id(current_book.get("bookId"))
            if current_book_id:
                next_current_books.append(
                    {
                        "bookId": current_book_id,
                        "currentPage": current_book.get("currentPage", 0),
                    }
                )
                continue

            book = ensure_book(books_collection, current_book)
            title_to_book_id[current_book.get("title")] = book["_id"]
            next_current_books.append(
                {
                    "bookId": book["_id"],
                    "currentPage": current_book.get("currentPage", 0),
                }
            )
            changed = True

        next_sessions = []
        for session in sessions:
            session_book_id = normalize_object_id(session.get("bookId"))
            if session_book_id:
                session["bookId"] = session_book_id
                next_sessions.append(session)
                continue

            book_id = title_to_book_id.get(session.get("bookTitle"))
            if not book_id:
                matching_book = next(
                    (book for book in current_books if book.get("title") == session.get("bookTitle")),
                    None,
                )
                book = ensure_book(
                    books_collection,
                    matching_book or {"title": session.get("bookTitle"), "pagesRead": session.get("pagesRead")},
                )
                book_id = book["_id"]

            next_sessions.append({**session, "bookId": book_id})
            changed = True

        if not changed:
            continue

        kids_collection.update_one(
            {"_id": kid["_id"]},
            {
                "$set": {
                    "reading.currentBooks": next_current_books,
                    "reading.sessions": next_sessions,
                }
            },
        )
        updated_kids += 1

    book_count = books_collection.count_documents({})
    print(f"Migrated {updated_kids} kid documents. Books collection now has {book_count} records.")


if __name__ == "__main__":
    try:
        main()
    except Exception as error:
        print(f"Failed to migrate books: {error}")
        raise
